File: api/Enkl.Api/Services/TaskService.cs
using System.Text.Json;
using Enkl.Api.Data;
using Enkl.Api.Domain.Entities;
using Enkl.Api.Dtos;
using Enkl.Api.Validation;
using Microsoft.EntityFrameworkCore;

namespace Enkl.Api.Services;

public class TaskService
{
    private readonly AppDbContext _db;

    public TaskService(AppDbContext db)
    {
        _db = db;
    }

    public async Task<TaskDto?> CreateAsync(Guid projectId, CreateTaskRequest request)
    {
        var project = await _db.Projects.FirstOrDefaultAsync(p => p.Id == projectId);
        var column = await _db.Columns.AsNoTracking().FirstOrDefaultAsync(c => c.Id == request.ColumnId && c.ProjectId == projectId);
        if (project is null || column is null) return null;

        if (request.DependsOnTaskIds is { Count: > 0 } && await WouldCreateDependencyCycleAsync(projectId, Guid.Empty, request.DependsOnTaskIds, isNewTask: true))
        {
            throw new ApiValidationException("That set of dependencies would create a cycle.");
        }
        if (request.ParentTaskId is { } newParentForCreate && await WouldCreateParentCycleAsync(projectId, Guid.Empty, newParentForCreate, isNewTask: true))
        {
            throw new ApiValidationException("That parent task would create a cycle.");
        }

        var now = DateTime.UtcNow;
        var task = new TaskItem
        {
            Id = Guid.NewGuid(),
            ProjectId = projectId,
            Key = $"{project.Key}-{project.TaskCounter}",
            Title = request.Title,
            Description = request.Description,
            Priority = request.Priority,
            ColumnId = request.ColumnId,
            AssigneeId = request.AssigneeId,
            ReleaseId = request.ReleaseId,
            TypeId = request.TypeId,
            ParentTaskId = request.ParentTaskId,
            DateCreated = now,
            DateLastModified = now,
            // A task created directly into a Done column counts as transitioning to Done immediately,
            // same rule as mutations.js's addTask (mutations.js:1056).
            DateDone = column.Done ? now : null,
            Progress = 0
        };
        project.TaskCounter++;

        _db.Tasks.Add(task);

        foreach (var depId in (request.DependsOnTaskIds ?? new List<Guid>()).Distinct())
        {
            if (depId != task.Id) _db.TaskDependencies.Add(new TaskDependency { TaskId = task.Id, DependsOnTaskId = depId });
        }

        await _db.SaveChangesAsync();
        return await ToTaskDtoWithRelationsAsync(task.Id);
    }

    public async Task<TaskDto?> UpdateAsync(Guid projectId, Guid taskId, UpdateTaskRequest request, string? changedByDisplayName)
    {
        var task = await _db.Tasks
            .Include(t => t.Dependencies)
            .FirstOrDefaultAsync(t => t.Id == taskId && t.ProjectId == projectId);
        if (task is null) return null;

        var newColumn = await _db.Columns.AsNoTracking().FirstOrDefaultAsync(c => c.Id == request.ColumnId && c.ProjectId == projectId);
        if (newColumn is null) return null;

        var newDeps = (request.DependsOnTaskIds ?? new List<Guid>()).Where(id => id != taskId).Distinct().ToList();
        if (await WouldCreateDependencyCycleAsync(projectId, taskId, newDeps, isNewTask: false))
        {
            throw new ApiValidationException("That set of dependencies would create a cycle.");
        }
        if (request.ParentTaskId is { } newParent && await WouldCreateParentCycleAsync(projectId, taskId, newParent, isNewTask: false))
        {
            throw new ApiValidationException("That parent task would create a cycle.");
        }

        var before = CaptureAuditSnapshot(task);
        var wasDone = task.DateDone.HasValue;
        var now = DateTime.UtcNow;

        task.Title = request.Title;
        task.Description = request.Description;
        task.Priority = request.Priority;
        task.ColumnId = request.ColumnId;
        task.AssigneeId = request.AssigneeId;
        task.ReleaseId = request.ReleaseId;
        task.TypeId = request.TypeId;
        task.ParentTaskId = request.ParentTaskId;
        task.DocumentationUrl = request.DocumentationUrl;
        task.StartDate = request.StartDate;
        task.EndDate = request.EndDate;
        task.BusinessValue = request.BusinessValue;
        task.TaskCost = request.TaskCost;
        task.Progress = request.Progress;
        task.EstimatedEffort = request.EstimatedEffort;
        task.ActualEffort = request.ActualEffort;
        task.Archived = request.Archived;
        task.DateLastModified = now;

        // Ported from mutations.js's updateTask (mutations.js:1198-1204): dateDone marks the most
        // recent time this task actually became Done, cleared if it leaves a Done column again.
        if (newColumn.Done && !wasDone) task.DateDone = now;
        else if (!newColumn.Done && wasDone) task.DateDone = null;

        _db.TaskDependencies.RemoveRange(task.Dependencies);
        foreach (var depId in newDeps)
        {
            _db.TaskDependencies.Add(new TaskDependency { TaskId = task.Id, DependsOnTaskId = depId });
        }

        if (await IsChangeAuditingEnabledAsync(projectId))
        {
            var after = CaptureAuditSnapshot(task, newDeps);
            RecordAuditEntries(task, before, after with { DependsOnTaskIds = newDeps }, changedByDisplayName);
        }

        await _db.SaveChangesAsync();
        return await ToTaskDtoWithRelationsAsync(task.Id);
    }

    /// <summary>
    /// Used by TasksController to capture a deleted task's key/title for its SSE broadcast *before*
    /// DeleteAsync removes the row — there's nothing left to read it from afterward.
    /// </summary>
    public async Task<(Guid TaskId, string Key, string Title)?> GetTaskSummaryAsync(Guid projectId, Guid taskId)
    {
        var task = await _db.Tasks.AsNoTracking().FirstOrDefaultAsync(t => t.Id == taskId && t.ProjectId == projectId);
        return task is null ? null : (task.Id, task.Key, task.Title);
    }

    public async Task<bool> DeleteAsync(Guid projectId, Guid taskId)
    {
        var task = await _db.Tasks.FirstOrDefaultAsync(t => t.Id == taskId && t.ProjectId == projectId);
        if (task is null) return false;

        // Mirrors mutations.js's deleteTask: a deleted task's sub-tasks are orphaned back to
        // top-level rather than cascade-deleted, and any Document/Risk/Decision pointing at it
        // is unlinked rather than blocked — matches the SetNull FKs already configured, but
        // ParentTaskId is Restrict so it needs the explicit unlink below.
        var children = await _db.Tasks.Where(t => t.ParentTaskId == taskId).ToListAsync();
        foreach (var child in children) child.ParentTaskId = null;

        _db.Tasks.Remove(task);
        await _db.SaveChangesAsync();
        return true;
    }

    /// <summary>Used by TasksController to know who to notify over SSE after a task change (see SseBroadcaster).</summary>
    public Task<List<Guid>> GetProjectMemberUserIdsAsync(Guid projectId) =>
        _db.ProjectMembers.Where(m => m.ProjectId == projectId).Select(m => m.UserId).ToListAsync();

    private async Task<TaskDto> ToTaskDtoWithRelationsAsync(Guid taskId)
    {
        var task = await _db.Tasks
            .AsNoTracking()
            .Include(t => t.Dependencies)
            .Include(t => t.AuditLog)
            .FirstAsync(t => t.Id == taskId);
        return ProjectService.ToTaskDto(task);
    }

    private async Task<bool> WouldCreateDependencyCycleAsync(Guid projectId, Guid taskId, List<Guid> newDeps, bool isNewTask)
    {
        var adjacency = await _db.TaskDependencies
            .Where(d => d.Task.ProjectId == projectId)
            .GroupBy(d => d.TaskId)
            .Select(g => new { g.Key, Deps = g.Select(x => x.DependsOnTaskId).ToList() })
            .ToDictionaryAsync(g => g.Key, g => g.Deps);

        var effectiveId = isNewTask ? Guid.NewGuid() : taskId;
        adjacency[effectiveId] = newDeps;
        return CycleDetection.HasCycle(adjacency);
    }

    private async Task<bool> WouldCreateParentCycleAsync(Guid projectId, Guid taskId, Guid newParentId, bool isNewTask)
    {
        var parentById = await _db.Tasks
            .Where(t => t.ProjectId == projectId)
            .Select(t => new { t.Id, t.ParentTaskId })
            .ToDictionaryAsync(t => t.Id, t => t.ParentTaskId);

        var effectiveId = isNewTask ? Guid.NewGuid() : taskId;
        parentById[effectiveId] = newParentId;
        return CycleDetection.HasParentCycle(parentById);
    }

    private async Task<bool> IsChangeAuditingEnabledAsync(Guid projectId)
    {
        var json = await _db.Projects.Where(p => p.Id == projectId).Select(p => p.HeaderButtonVisibilityJson).FirstOrDefaultAsync();
        if (string.IsNullOrEmpty(json)) return false;
        try
        {
            using var doc = JsonDocument.Parse(json);
            return doc.RootElement.TryGetProperty("changeAuditing", out var val) && val.ValueKind == JsonValueKind.True;
        }
        catch (JsonException)
        {
            return false;
        }
    }

    private static readonly (string Field, Func<TaskAuditSnapshot, object?> Get)[] AuditDiffedFields =
    {
        ("title", s => s.Title), ("description", s => s.Description), ("priority", s => s.Priority),
        ("assigneeId", s => s.AssigneeId), ("releaseId", s => s.ReleaseId), ("typeId", s => s.TypeId),
        ("documentationUrl", s => s.DocumentationUrl), ("startDate", s => s.StartDate), ("endDate", s => s.EndDate),
        ("businessValue", s => s.BusinessValue), ("taskCost", s => s.TaskCost), ("progress", s => s.Progress),
        ("estimatedEffort", s => s.EstimatedEffort), ("actualEffort", s => s.ActualEffort),
        ("archived", s => s.Archived), ("dependencies", s => s.DependsOnTaskIds), ("parentTaskId", s => s.ParentTaskId)
    };

    private static TaskAuditSnapshot CaptureAuditSnapshot(TaskItem t, List<Guid>? deps = null) => new(
        t.Title, t.Description, t.Priority, t.AssigneeId, t.ReleaseId, t.TypeId, t.DocumentationUrl,
        t.StartDate, t.EndDate, t.BusinessValue, t.TaskCost, t.Progress, t.EstimatedEffort, t.ActualEffort,
        t.Archived, deps ?? t.Dependencies.Select(d => d.DependsOnTaskId).ToList(), t.ParentTaskId);

    private void RecordAuditEntries(TaskItem task, TaskAuditSnapshot before, TaskAuditSnapshot after, string? changedBy)
    {
        var now = DateTime.UtcNow;
        foreach (var (field, get) in AuditDiffedFields)
        {
            var oldVal = get(before);
            var newVal = get(after);
            if (!AuditValuesEqual(oldVal, newVal))
            {
                _db.TaskAuditLogEntries.Add(new TaskAuditLogEntry
                {
                    Id = Guid.NewGuid(), TaskId = task.Id, Timestamp = now, Field = field,
                    OldValue = FormatAuditValue(oldVal), NewValue = FormatAuditValue(newVal), ChangedBy = changedBy
                });
            }
        }
    }

    private static bool AuditValuesEqual(object? a, object? b)
    {
        if (a is List<Guid> la && b is List<Guid> lb)
        {
            return la.OrderBy(x => x).SequenceEqual(lb.OrderBy(x => x));
        }
        return Equals(a, b);
    }

    private static string? FormatAuditValue(object? value) => value switch
    {
        null => null,
        List<Guid> list => list.Count == 0 ? "[]" : string.Join(",", list),
        _ => value.ToString()
    };

    private record TaskAuditSnapshot(
        string Title, string? Description, string Priority, Guid? AssigneeId, Guid? ReleaseId, Guid? TypeId,
        string? DocumentationUrl, DateOnly? StartDate, DateOnly? EndDate, int? BusinessValue, int? TaskCost,
        int Progress, decimal? EstimatedEffort, decimal? ActualEffort, bool Archived, List<Guid> DependsOnTaskIds,
        Guid? ParentTaskId);
}

File: mariadb-api/src/Services/TaskService.php
<?php

declare(strict_types=1);

namespace Enkl\Api\Services;

use Enkl\Api\Support\SqlDateTime;
use Enkl\Api\Support\Uuid;
use Enkl\Api\Validation\ApiValidationException;
use Enkl\Api\Validation\CycleDetection;
use PDO;

/** Ported from Services/TaskService.cs — see that file's own comments for the "why" behind each piece; kept here too where it isn't obvious from the code alone. */
final class TaskService
{
    private const AUDIT_DIFFED_FIELDS = [
        'title', 'description', 'priority', 'assigneeId', 'releaseId', 'typeId', 'documentationUrl',
        'startDate', 'endDate', 'businessValue', 'taskCost', 'progress', 'estimatedEffort',
        'actualEffort', 'archived', 'dependencies', 'parentTaskId',
    ];

    public function __construct(private readonly PDO $db)
    {
    }

    // ARCHITECTURE-REVIEW.md finding 3.1: the Task INSERT, the Projects.TaskCounter increment, and
    // the TaskDependencies INSERTs used to be separately auto-committed — a failure after the Task
    // insert but before the counter increment leaves the NEXT created task in this project reusing
    // the same Key (duplicate task keys, likely a confusing unique-constraint failure on the next
    // create instead of this one); a failure mid-dependency-loop leaves a task created with an
    // incomplete dependency list and no error surfaced about it.
    public function create(string $projectId, array $request): ?array
    {
        $this->db->beginTransaction();
        try {
            $result = $this->createInTransaction($projectId, $request);
            $this->db->commit();
            return $result;
        } catch (\Throwable $e) {
            if ($this->db->inTransaction()) {
                $this->db->rollBack();
            }
            throw $e;
        }
    }

    private function createInTransaction(string $projectId, array $request): ?array
    {
        $stmt = $this->db->prepare('SELECT * FROM "Projects" WHERE "Id" = :id');
        $stmt->execute(['id' => $projectId]);
        $project = $stmt->fetch();

        $stmt = $this->db->prepare('SELECT * FROM "Columns" WHERE "Id" = :id AND "ProjectId" = :pid');
        $stmt->execute(['id' => $request['columnId'] ?? null, 'pid' => $projectId]);
        $column = $stmt->fetch();

        if ($project === false || $column === false) {
            return null;
        }

        $dependsOn = array_values(array_unique($request['dependsOnTaskIds'] ?? []));
        if ($dependsOn !== [] && $this->wouldCreateDependencyCycle($projectId, null, $dependsOn)) {
            throw new ApiValidationException('That set of dependencies would create a cycle.');
        }
        $parentTaskId = $request['parentTaskId'] ?? null;
        if ($parentTaskId !== null && $this->wouldCreateParentCycle($projectId, null, $parentTaskId)) {
            throw new ApiValidationException('That parent task would create a cycle.');
        }

        $taskId = Uuid::v4();
        $key = $project['Key'] . '-' . $project['TaskCounter'];
        $done = (bool) $column['Done'];

        $stmt = $this->db->prepare(<<<SQL
            INSERT INTO "Tasks" (
                "Id", "ProjectId", "Key", "Title", "Description", "Priority", "ColumnId", "AssigneeId",
                "ReleaseId", "TypeId", "ParentTaskId", "DocumentationUrl", "StartDate", "EndDate",
                "BusinessValue", "TaskCost", "EstimatedEffort", "ActualEffort", "Archived",
                "DateCreated", "DateLastModified", "DateDone", "Progress"
            ) VALUES (
                :id, :pid, :key, :title, :description, :priority, :columnId, :assigneeId,
                :releaseId, :typeId, :parentTaskId, :documentationUrl, :startDate, :endDate,
                :businessValue, :taskCost, :estimatedEffort, :actualEffort, :archived,
                now(), now(), :dateDone, :progress
            )
        SQL);
        $stmt->execute([
            'id' => $taskId, 'pid' => $projectId, 'key' => $key,
            'title' => $request['title'] ?? '', 'description' => $request['description'] ?? null,
            'priority' => $request['priority'] ?? 'medium', 'columnId' => $request['columnId'],
            'assigneeId' => $request['assigneeId'] ?? null, 'releaseId' => $request['releaseId'] ?? null,
            'typeId' => $request['typeId'] ?? null, 'parentTaskId' => $parentTaskId,
            'documentationUrl' => $request['documentationUrl'] ?? null,
            'startDate' => $request['startDate'] ?? null, 'endDate' => $request['endDate'] ?? null,
            'businessValue' => $request['businessValue'] ?? null, 'taskCost' => $request['taskCost'] ?? null,
            'estimatedEffort' => $request['estimatedEffort'] ?? null, 'actualEffort' => $request['actualEffort'] ?? null,
            'archived' => (int) (bool) ($request['archived'] ?? false),
            'progress' => (int) ($request['progress'] ?? 0),
            'dateDone' => $done ? SqlDateTime::now() : null,
        ]);

        $this->db->prepare('UPDATE "Projects" SET "TaskCounter" = "TaskCounter" + 1 WHERE "Id" = :id')->execute(['id' => $projectId]);

        $depStmt = $this->db->prepare('INSERT INTO "TaskDependencies" ("TaskId", "DependsOnTaskId") VALUES (:tid, :did)');
        foreach ($dependsOn as $depId) {
            if ($depId !== $taskId) {
                $depStmt->execute(['tid' => $taskId, 'did' => $depId]);
            }
        }

        return $this->toTaskDto($taskId);
    }

    // ARCHITECTURE-REVIEW.md finding 3.1: the Task UPDATE, the TaskDependencies DELETE+re-INSERT, and
    // recordAuditEntries()'s own INSERTs used to be separately auto-committed — a failure mid-sequence
    // left the task's fields updated but its dependency list half-cleared, or audit history missing
    // for a change that did actually take effect.
    public function update(string $projectId, string $taskId, array $request, ?string $changedByDisplayName): ?array
    {
        $this->db->beginTransaction();
        try {
            $result = $this->updateInTransaction($projectId, $taskId, $request, $changedByDisplayName);
            $this->db->commit();
            return $result;
        } catch (\Throwable $e) {
            if ($this->db->inTransaction()) {
                $this->db->rollBack();
            }
            throw $e;
        }
    }

    private function updateInTransaction(string $projectId, string $taskId, array $request, ?string $changedByDisplayName): ?array
    {
        $stmt = $this->db->prepare('SELECT * FROM "Tasks" WHERE "Id" = :id AND "ProjectId" = :pid');
        $stmt->execute(['id' => $taskId, 'pid' => $projectId]);
        $task = $stmt->fetch();
        if ($task === false) {
            return null;
        }

        $stmt = $this->db->prepare('SELECT * FROM "Columns" WHERE "Id" = :id AND "ProjectId" = :pid');
        $stmt->execute(['id' => $request['columnId'] ?? null, 'pid' => $projectId]);
        $newColumn = $stmt->fetch();
        if ($newColumn === false) {
            return null;
        }

        $newDeps = array_values(array_unique(array_filter(
            $request['dependsOnTaskIds'] ?? [],
            static fn($id) => $id !== $taskId
        )));
        if ($this->wouldCreateDependencyCycle($projectId, $taskId, $newDeps)) {
            throw new ApiValidationException('That set of dependencies would create a cycle.');
        }
        $newParentId = $request['parentTaskId'] ?? null;
        if ($newParentId !== null && $this->wouldCreateParentCycle($projectId, $taskId, $newParentId)) {
            throw new ApiValidationException('That parent task would create a cycle.');
        }

        $before = $this->captureAuditSnapshot($task, $this->currentDependencyIds($taskId));
        $wasDone = $task['DateDone'] !== null;

        $newDateDone = $task['DateDone'];
        if ((bool) $newColumn['Done'] && !$wasDone) {
            $newDateDone = SqlDateTime::now();
        } elseif (!$newColumn['Done'] && $wasDone) {
            $newDateDone = null;
        }

        $stmt = $this->db->prepare(<<<SQL
            UPDATE "Tasks" SET
                "Title" = :title, "Description" = :description, "Priority" = :priority, "ColumnId" = :columnId,
                "AssigneeId" = :assigneeId, "ReleaseId" = :releaseId, "TypeId" = :typeId, "ParentTaskId" = :parentTaskId,
                "DocumentationUrl" = :documentationUrl, "StartDate" = :startDate, "EndDate" = :endDate,
                "BusinessValue" = :businessValue, "TaskCost" = :taskCost, "Progress" = :progress,
                "EstimatedEffort" = :estimatedEffort, "ActualEffort" = :actualEffort, "Archived" = :archived,
                "DateLastModified" = now(), "DateDone" = :dateDone
            WHERE "Id" = :id
        SQL);
        $stmt->execute([
            'title' => $request['title'] ?? '', 'description' => $request['description'] ?? null,
            'priority' => $request['priority'] ?? 'medium', 'columnId' => $request['columnId'],
            'assigneeId' => $request['assigneeId'] ?? null, 'releaseId' => $request['releaseId'] ?? null,
            'typeId' => $request['typeId'] ?? null, 'parentTaskId' => $newParentId,
            'documentationUrl' => $request['documentationUrl'] ?? null,
            'startDate' => $request['startDate'] ?? null, 'endDate' => $request['endDate'] ?? null,
            'businessValue' => $request['businessValue'] ?? null, 'taskCost' => $request['taskCost'] ?? null,
            'progress' => (int) ($request['progress'] ?? 0),
            'estimatedEffort' => $request['estimatedEffort'] ?? null, 'actualEffort' => $request['actualEffort'] ?? null,
            // (int), not the raw PHP bool — PDO's array-form execute() would bind false as '' otherwise,
            // which Postgres's boolean parser rejects.
            'archived' => (int) (bool) ($request['archived'] ?? false),
            'dateDone' => $newDateDone, 'id' => $taskId,
        ]);

        $this->db->prepare('DELETE FROM "TaskDependencies" WHERE "TaskId" = :id')->execute(['id' => $taskId]);
        $depStmt = $this->db->prepare('INSERT INTO "TaskDependencies" ("TaskId", "DependsOnTaskId") VALUES (:tid, :did)');
        foreach ($newDeps as $depId) {
            $depStmt->execute(['tid' => $taskId, 'did' => $depId]);
        }

        if ($this->isChangeAuditingEnabled($projectId)) {
            $updatedTask = array_merge($task, [
                'Title' => $request['title'] ?? '', 'Description' => $request['description'] ?? null,
                'Priority' => $request['priority'] ?? 'medium', 'AssigneeId' => $request['assigneeId'] ?? null,
                'ReleaseId' => $request['releaseId'] ?? null, 'TypeId' => $request['typeId'] ?? null,
                'DocumentationUrl' => $request['documentationUrl'] ?? null,
                'StartDate' => $request['startDate'] ?? null, 'EndDate' => $request['endDate'] ?? null,
                'BusinessValue' => $request['businessValue'] ?? null, 'TaskCost' => $request['taskCost'] ?? null,
                'Progress' => (int) ($request['progress'] ?? 0),
                'EstimatedEffort' => $request['estimatedEffort'] ?? null, 'ActualEffort' => $request['actualEffort'] ?? null,
                'Archived' => (bool) ($request['archived'] ?? false), 'ParentTaskId' => $newParentId,
            ]);
            $after = $this->captureAuditSnapshot($updatedTask, $newDeps);
            $this->recordAuditEntries($taskId, $before, $after, $changedByDisplayName);
        }

        return $this->toTaskDto($taskId);
    }

    public function getTaskSummary(string $projectId, string $taskId): ?array
    {
        $stmt = $this->db->prepare('SELECT "Id", "Key", "Title" FROM "Tasks" WHERE "Id" = :id AND "ProjectId" = :pid');
        $stmt->execute(['id' => $taskId, 'pid' => $projectId]);
        $row = $stmt->fetch();
        return $row === false ? null : ['taskId' => $row['Id'], 'key' => $row['Key'], 'title' => $row['Title']];
    }

    // ARCHITECTURE-REVIEW.md finding 3.1: orphaning sub-tasks and deleting the task used to be two
    // separately auto-committed statements — a failure on the DELETE left sub-tasks already
    // unlinked from a parent that still exists.
    public function delete(string $projectId, string $taskId): bool
    {
        $this->db->beginTransaction();
        try {
            $result = $this->deleteInTransaction($projectId, $taskId);
            $this->db->commit();
            return $result;
        } catch (\Throwable $e) {
            if ($this->db->inTransaction()) {
                $this->db->rollBack();
            }
            throw $e;
        }
    }

    private function deleteInTransaction(string $projectId, string $taskId): bool
    {
        $stmt = $this->db->prepare('SELECT 1 FROM "Tasks" WHERE "Id" = :id AND "ProjectId" = :pid');
        $stmt->execute(['id' => $taskId, 'pid' => $projectId]);
        if ($stmt->fetch() === false) {
            return false;
        }

        // ParentTaskId is a Restrict FK (see TaskItemConfiguration) — sub-tasks are orphaned back to
        // top-level, not cascade-deleted, mirroring mutations.js's deleteTask.
        $this->db->prepare('UPDATE "Tasks" SET "ParentTaskId" = NULL WHERE "ParentTaskId" = :id')->execute(['id' => $taskId]);
        $this->db->prepare('DELETE FROM "Tasks" WHERE "Id" = :id')->execute(['id' => $taskId]);
        return true;
    }

    public function getProjectMemberUserIds(string $projectId): array
    {
        $stmt = $this->db->prepare('SELECT "UserId" FROM "ProjectMembers" WHERE "ProjectId" = :pid');
        $stmt->execute(['pid' => $projectId]);
        return $stmt->fetchAll(PDO::FETCH_COLUMN);
    }

    /** Shared with ProjectService::getProjectDetail — builds every task DTO (with dependencies + audit log + comments) for a project in one pass. */
    public static function fetchTaskDtos(PDO $db, string $projectId): array
    {
        $stmt = $db->prepare('SELECT * FROM "Tasks" WHERE "ProjectId" = :pid');
        $stmt->execute(['pid' => $projectId]);
        $tasks = $stmt->fetchAll();

        $depStmt = $db->prepare('SELECT "DependsOnTaskId" FROM "TaskDependencies" WHERE "TaskId" = :id');
        $auditStmt = $db->prepare('SELECT * FROM "TaskAuditLogEntries" WHERE "TaskId" = :id ORDER BY "Timestamp" ASC');
        $commentStmt = $db->prepare('SELECT * FROM "TaskComments" WHERE "TaskId" = :id ORDER BY "DateCreated" ASC');

        return array_map(function (array $t) use ($depStmt, $auditStmt, $commentStmt): array {
            $depStmt->execute(['id' => $t['Id']]);
            $auditStmt->execute(['id' => $t['Id']]);
            $commentStmt->execute(['id' => $t['Id']]);
            return self::mapTaskRow($t, $depStmt->fetchAll(PDO::FETCH_COLUMN), array_map(
                static fn(array $a): array => [
                    'id' => $a['Id'], 'timestamp' => $a['Timestamp'], 'field' => $a['Field'],
                    'oldValue' => $a['OldValue'], 'newValue' => $a['NewValue'], 'changedBy' => $a['ChangedBy'],
                ],
                $auditStmt->fetchAll()
            ), self::mapComments($commentStmt->fetchAll()));
        }, $tasks);
    }

    private function toTaskDto(string $taskId): array
    {
        $stmt = $this->db->prepare('SELECT * FROM "Tasks" WHERE "Id" = :id');
        $stmt->execute(['id' => $taskId]);
        $task = $stmt->fetch();

        $depStmt = $this->db->prepare('SELECT "DependsOnTaskId" FROM "TaskDependencies" WHERE "TaskId" = :id');
        $depStmt->execute(['id' => $taskId]);

        $auditStmt = $this->db->prepare('SELECT * FROM "TaskAuditLogEntries" WHERE "TaskId" = :id ORDER BY "Timestamp" ASC');
        $auditStmt->execute(['id' => $taskId]);

        $commentStmt = $this->db->prepare('SELECT * FROM "TaskComments" WHERE "TaskId" = :id ORDER BY "DateCreated" ASC');
        $commentStmt->execute(['id' => $taskId]);

        return self::mapTaskRow($task, $depStmt->fetchAll(PDO::FETCH_COLUMN), array_map(
            static fn(array $a): array => [
                'id' => $a['Id'], 'timestamp' => $a['Timestamp'], 'field' => $a['Field'],
                'oldValue' => $a['OldValue'], 'newValue' => $a['NewValue'], 'changedBy' => $a['ChangedBy'],
            ],
            $auditStmt->fetchAll()
        ), self::mapComments($commentStmt->fetchAll()));
    }

    /** @param array<int,array<string,mixed>> $rows */
    private static function mapComments(array $rows): array
    {
        return array_map(static fn(array $c): array => [
            'id' => $c['Id'], 'text' => $c['Text'], 'dateCreated' => $c['DateCreated'],
            'authorId' => $c['AuthorId'], 'authorName' => $c['AuthorName'],
        ], $rows);
    }

    private static function mapTaskRow(array $t, array $dependsOnTaskIds, array $auditLog, array $comments = []): array
    {
        return [
            'id' => $t['Id'], 'key' => $t['Key'], 'title' => $t['Title'], 'description' => $t['Description'],
            'priority' => $t['Priority'], 'columnId' => $t['ColumnId'], 'assigneeId' => $t['AssigneeId'],
            'releaseId' => $t['ReleaseId'], 'typeId' => $t['TypeId'], 'parentTaskId' => $t['ParentTaskId'],
            'documentationUrl' => $t['DocumentationUrl'], 'dateCreated' => $t['DateCreated'],
            'dateLastModified' => $t['DateLastModified'], 'dateDone' => $t['DateDone'],
            'startDate' => $t['StartDate'], 'endDate' => $t['EndDate'],
            'businessValue' => $t['BusinessValue'] !== null ? (int) $t['BusinessValue'] : null,
            'taskCost' => $t['TaskCost'] !== null ? (int) $t['TaskCost'] : null,
            'progress' => (int) $t['Progress'],
            'estimatedEffort' => $t['EstimatedEffort'] !== null ? (float) $t['EstimatedEffort'] : null,
            'actualEffort' => $t['ActualEffort'] !== null ? (float) $t['ActualEffort'] : null,
            'archived' => (bool) $t['Archived'],
            'dependsOnTaskIds' => $dependsOnTaskIds,
            'auditLog' => $auditLog,
            'comments' => $comments,
        ];
    }

    private function currentDependencyIds(string $taskId): array
    {
        $stmt = $this->db->prepare('SELECT "DependsOnTaskId" FROM "TaskDependencies" WHERE "TaskId" = :id');
        $stmt->execute(['id' => $taskId]);
        return $stmt->fetchAll(PDO::FETCH_COLUMN);
    }

    private function wouldCreateDependencyCycle(string $projectId, ?string $taskId, array $newDeps): bool
    {
        $stmt = $this->db->prepare(<<<SQL
            SELECT d."TaskId", d."DependsOnTaskId" FROM "TaskDependencies" d
            JOIN "Tasks" t ON t."Id" = d."TaskId"
            WHERE t."ProjectId" = :pid
        SQL);
        $stmt->execute(['pid' => $projectId]);

        $adjacency = [];
        foreach ($stmt->fetchAll() as $row) {
            $adjacency[$row['TaskId']][] = $row['DependsOnTaskId'];
        }

        $effectiveId = $taskId ?? Uuid::v4();
        $adjacency[$effectiveId] = $newDeps;

        return CycleDetection::hasCycle($adjacency);
    }

    private function wouldCreateParentCycle(string $projectId, ?string $taskId, string $newParentId): bool
    {
        $stmt = $this->db->prepare('SELECT "Id", "ParentTaskId" FROM "Tasks" WHERE "ProjectId" = :pid');
        $stmt->execute(['pid' => $projectId]);

        $parentById = [];
        foreach ($stmt->fetchAll() as $row) {
            $parentById[$row['Id']] = $row['ParentTaskId'];
        }

        $effectiveId = $taskId ?? Uuid::v4();
        $parentById[$effectiveId] = $newParentId;

        return CycleDetection::hasParentCycle($parentById);
    }

    private function isChangeAuditingEnabled(string $projectId): bool
    {
        $stmt = $this->db->prepare('SELECT "HeaderButtonVisibilityJson" FROM "Projects" WHERE "Id" = :id');
        $stmt->execute(['id' => $projectId]);
        $json = $stmt->fetchColumn();
        if (!$json) {
            return false;
        }
        $decoded = json_decode($json, true);
        return is_array($decoded) && ($decoded['changeAuditing'] ?? false) === true;
    }

    private function captureAuditSnapshot(array $task, array $dependsOnTaskIds): array
    {
        return [
            'title' => $task['Title'], 'description' => $task['Description'], 'priority' => $task['Priority'],
            'assigneeId' => $task['AssigneeId'], 'releaseId' => $task['ReleaseId'], 'typeId' => $task['TypeId'],
            'documentationUrl' => $task['DocumentationUrl'], 'startDate' => $task['StartDate'], 'endDate' => $task['EndDate'],
            'businessValue' => $task['BusinessValue'], 'taskCost' => $task['TaskCost'], 'progress' => (int) $task['Progress'],
            'estimatedEffort' => $task['EstimatedEffort'], 'actualEffort' => $task['ActualEffort'],
            'archived' => (bool) $task['Archived'], 'dependencies' => $dependsOnTaskIds, 'parentTaskId' => $task['ParentTaskId'],
        ];
    }

    private function recordAuditEntries(string $taskId, array $before, array $after, ?string $changedBy): void
    {
        $stmt = $this->db->prepare(<<<SQL
            INSERT INTO "TaskAuditLogEntries" ("Id", "TaskId", "Timestamp", "Field", "OldValue", "NewValue", "ChangedBy")
            VALUES (:id, :taskId, now(), :field, :oldValue, :newValue, :changedBy)
        SQL);

        foreach (self::AUDIT_DIFFED_FIELDS as $field) {
            $oldVal = $before[$field];
            $newVal = $after[$field];
            if (!$this->auditValuesEqual($oldVal, $newVal)) {
                $stmt->execute([
                    'id' => Uuid::v4(), 'taskId' => $taskId, 'field' => $field,
                    'oldValue' => $this->formatAuditValue($oldVal), 'newValue' => $this->formatAuditValue($newVal),
                    'changedBy' => $changedBy,
                ]);
            }
        }
    }

    private function auditValuesEqual(mixed $a, mixed $b): bool
    {
        if (is_array($a) && is_array($b)) {
            $sa = $a;
            $sb = $b;
            sort($sa);
            sort($sb);
            return $sa === $sb;
        }
        // Loose-ish but type-safe: bools/ints/strings/null compare by value, matching C#'s Equals()
        // for the corresponding boxed primitive types used in TaskAuditSnapshot.
        return $a === $b || (is_numeric($a) && is_numeric($b) && (float) $a === (float) $b);
    }

    private function formatAuditValue(mixed $value): ?string
    {
        if ($value === null) {
            return null;
        }
        if (is_array($value)) {
            return $value === [] ? '[]' : implode(',', $value);
        }
        if (is_bool($value)) {
            return $value ? 'True' : 'False'; // matches C#'s bool.ToString()
        }
        return (string) $value;
    }
}
